singularity-eos/eos/modifiers/ramps_eos.hpp

```diff
@@ -0,0 +1,213 @@
+
+  auto GetOnDevice() { return SAPRampEOS<T>(t_.GetOnDevice(), r0_, a_, b_, c_); }
+  inline void Finalize() { t_.Finalize(); }
+
+  PORTABLE_INLINE_FUNCTION
+  Real get_ramp_pressure(Real rho) const {
+    const Real p_ramp{rho < r0_     ? 0.0
+                      : rho < rmid_ ? a_ * (rho / r0_ - 1.0)
+                                    : b_ * (rho / r0_ - c_)};
+    return p_ramp;
+  }
+
+  PORTABLE_FUNCTION
+  Real TemperatureFromDensityInternalEnergy(const Real rho, const Real sie,
+                                            Real *lambda = nullptr) const {
+    return t_.TemperatureFromDensityInternalEnergy(rho, sie, lambda);
+  }
+  PORTABLE_FUNCTION
+  Real InternalEnergyFromDensityTemperature(const Real rho, const Real temperature,
+                                            Real *lambda = nullptr) const {
+    return t_.InternalEnergyFromDensityTemperature(rho, temperature, lambda);
+  }
+  PORTABLE_FUNCTION
+  Real PressureFromDensityInternalEnergy(const Real rho, const Real sie,
+                                         Real *lambda = nullptr) const {
+    // ramp pressure
+    const Real p_ramp{get_ramp_pressure(rho)};
+    // pressure from eos
+    const Real p_eos{t_.PressureFromDensityInternalEnergy(rho, sie, lambda)};
+    // return max(p_ramp, p_eos)
+    return p_eos < p_ramp ? p_ramp : p_eos;
+  }
+  PORTABLE_FUNCTION
+  Real SpecificHeatFromDensityInternalEnergy(const Real rho, const Real sie,
+                                             Real *lambda = nullptr) const {
+    return t_.SpecificHeatFromDensityInternalEnergy(rho, sie, lambda);
+  }
+  PORTABLE_FUNCTION
+  Real BulkModulusFromDensityInternalEnergy(const Real rho, const Real sie,
+                                            Real *lambda = nullptr) const {
+    return t_.BulkModulusFromDensityInternalEnergy(rho, sie, lambda);
+  }
+  PORTABLE_FUNCTION
+  Real GruneisenParamFromDensityInternalEnergy(const Real rho, const Real sie,
+                                               Real *lambda = nullptr) const {
+    return t_.GruneisenParamFromDensityInternalEnergy(rho, sie, lambda);
+  }
+  PORTABLE_FUNCTION
+  Real PressureFromDensityTemperature(const Real rho, const Real temperature,
+                                      Real *lambda = nullptr) const {
+    const Real p_ramp{get_ramp_pressure(rho)};
+    const Real p_eos{t_.PressureFromDensityTemperature(rho, temperature, lambda)};
+    return p_eos < p_ramp ? p_ramp : p_eos;
+  }
+  PORTABLE_FUNCTION
+  Real SpecificHeatFromDensityTemperature(const Real rho, const Real temperature,
+                                          Real *lambda = nullptr) const {
+    return t_.SpecificHeatFromDensityTemperature(rho, temperature, lambda);
+  }
+  PORTABLE_FUNCTION
+  Real BulkModulusFromDensityTemperature(const Real rho, const Real temperature,
+                                         Real *lambda = nullptr) const {
+    return t_.BulkModulusFromDensityTemperature(rho, temperature, lambda);
```

Comment:
I _believe_ most implementations of a ramp utilize the slope of the ramp itself in the bulk modulus calculation

Reply:
how do I tell when to use the ramp slope vs the eos evaluated one?

Reply:
You probably need to do the ramp pressure vs EOS pressure test to figure out if you're on the ramp or not

Reply:
To add to this, most ramps assume that the ramp defines a roughly isentropic path for the material which makes the slope of the ramp a direct input (i.e. no thermodynamic relations needed) to the isentropic bulk modulus calculation.

Whether this is physical or not is another question, but if we want to match the current implementations I would just use the slope of the ramp.